### DISCUSSION TITLE
Extract Text from Annotiations and corresponding highlighted Text

### DISCUSSION THREAD

**Discussion by M-M99** [pymupdf/PyMuPDF#1671] [Q&A] [answered]
Hello, 

can somebody explain me how i can get the Text which is highlighted in the PDF. 

I know, that i can get the annot Text with a loop over page.annot()  / annot.getText. 
But how i can get the corresponding Text which is highlighted on the PDF. 

Thanks!

**Comment by JorjMcKie** [comment, marked as answer]
There is one thing you must keep in mind:
Annotations **_are not part_** of the page's contents. Imagine them like dust on a nice painting on the wall. The items shown in the painting are not aware of any dust that may cover them. And like dust, annotations can be wiped out without changing the page itself.
You get the idea.
Accordingly, an annotation may cover just anything: text, drawings, images ... or nothing.
There is a rectangle associated with an annotation, `annot.rect`, which can be used to find out what is underneath it. For example do this to find any covered text: `text = page.get_text(clip=annot.rect)`.

But of course, highlight annotations (like their friends: underlines, strike-throughs, squiggles) **_are_** mostly used to mark text. So you might want to do this:
```python
for annot in page.annots(types=(fitz.PDF_ANNOT_HIGHLIGHT, fitz.PDF_ANNOT_UNDERLINE)):
    print(page.get_text(clip=annot.rect))
```

The annotation method `annot.get_text()` only delivers the text stored as part of some specific annotation types: freetext and stamps. Like mentioned above, this text is not part of the page, but part of the annotation.

**Reply by M-M99** [reply]
Hello JorjMcKie, 
thanks for your detailed answer! 
Helped me a lot!

**Reply by tristone13th** [reply]
Hello, sometimes we will add annotation text on the highlight annotations (maybe we can call it 'comment' on anl annotation). Could you show me how to.extract those text? thanks!

**Reply by JorjMcKie** [reply]
> Hello, sometimes we will add annotation text on the highlight annotations (maybe we can call it 'comment' on anl annotation). Could you show me how to.extract those text? thanks!

You mean popup annotations? If an annotation has a popup, the popup text is also stored in the **annotations** ``info["content"]``.

**Reply by tristone13th** [reply]
> 

It really helped, many thanks!

**Comment by JorjMcKie** [comment]
BTW: your post is a typical **_Discussions_** item, not an issue.

**Comment by sunic23** [comment]
Hi Jorj,
I am trying to create freeText callout annotation in 1.20.1 PyMupdf but not able to get it working and quite not clear on the doucment type[](https://pymupdf.readthedocs.io/en/latest/annot.html?highlight=callout#Annot.type)
A number and one or two strings describing the annotation type, like [2, ‘FreeText’, ‘FreeTextCallout’]. The second string entry is optional and may be empty. See the appendix [Annotation Types](https://pymupdf.readthedocs.io/en/latest/vars.html#annotationtypes) for a list of possible values and their meanings.

Could you please help in provide details on how to create a FreeText Cllaout Annotation ? 
![FreeTextCallout](https://user-images.githubusercontent.com/8039930/191930070-7d2812d0-505d-4a0c-aa2e-94f5250f8b3c.png)

**Reply by JorjMcKie** [reply]
If there exists a callout annot, PyMuPDF is able to report it. This is what you have found in the documentation.

But there currently is no support to **_create_** a callout annotation, sorry.

**Comment by tofDou** [comment]
I found something strange when dealing with extracting comments from PDF (made with Acrobat).
The annot.info["content"] is "empty" (not relevant in fact) when the state (Accepted, Canceled, Closed) of the comment is modified. Even if the state is reset to "none", the content text is no more available the content entry
